src/System.Memory/src/System/SpanHelpers.cs

```diff
@@ -106,6 +106,83 @@ private static bool IsReferenceFreeCoreSlow(Type type)
             return true;
         }
 
+        [StructLayout(LayoutKind.Sequential, Size = 64)]
+        private struct Reg64 { }
+        [StructLayout(LayoutKind.Sequential, Size = 32)]
+        private struct Reg32 { }
+        [StructLayout(LayoutKind.Sequential, Size = 16)]
+        private struct Reg16 { }
+
+        unsafe static readonly UIntPtr UIntPtrMask4 = sizeof(UIntPtr) == sizeof(uint) ? new UIntPtr(~3u) : new UIntPtr(~((ulong)(3u)));
+        unsafe static readonly UIntPtr UIntPtrMask8 = sizeof(UIntPtr) == sizeof(uint) ? new UIntPtr(~7u) : new UIntPtr(~((ulong)(7u)));
+        unsafe static readonly UIntPtr UIntPtrMask16 = sizeof(UIntPtr) == sizeof(uint) ? new UIntPtr(~15u) : new UIntPtr(~((ulong)(15u)));
+        unsafe static readonly UIntPtr UIntPtrMask32 = sizeof(UIntPtr) == sizeof(uint) ? new UIntPtr(~31u) : new UIntPtr(~((ulong)(31u)));
+        unsafe static readonly UIntPtr UIntPtrMask64 = sizeof(UIntPtr) == sizeof(uint) ? new UIntPtr(~63u) : new UIntPtr(~((ulong)(63u)));
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        unsafe static UIntPtr BitwiseAnd(this UIntPtr value, UIntPtr mask)
+        {
+            return (sizeof(UIntPtr) == sizeof(uint)) 
+                ? new UIntPtr((uint)value & (uint)mask)
+                : new UIntPtr((ulong)value & (ulong)mask);
+        }
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        unsafe static bool LessThan(this IntPtr index, UIntPtr length)
+        {
+            return (sizeof(UIntPtr) == sizeof(uint))
+                ? (uint)index < (uint)length
+                : (ulong)index < (ulong)length;
+        }
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        unsafe static bool LessThanEqual(this IntPtr index, UIntPtr length)
+        {
+            return (sizeof(UIntPtr) == sizeof(uint))
+                ? (int)index <= (int)length
+                : (long)index <= (long)length;
+        }
+
+        public unsafe static void ClearPointerSized(ref byte b, UIntPtr byteLength)
+        {
+            // TODO: Perhaps do switch casing... 
+            // TODO: Bitwise masking generates weird assembly 64-bit, including not inlining calls.
+
+            var i = IntPtr.Zero;
+            //while (i.LessThan(byteLength.BitwiseAnd(UIntPtrMask64)))
+            while (i.LessThanEqual(byteLength - sizeof(Reg64)))
+            {
+                Unsafe.As<byte, Reg64>(ref Unsafe.Add<byte>(ref b, i)) = default(Reg64);
```

Comment:
> Would IntPtrs4 also not be atomic?

I do not think that any spec says that this has to be atomic. It probably happens to be atomic most of the time, but hard to tell for sure.

> struct Objects4
> Would that then be guaranteed to be atomic?

The assignment of the individual elements should be atomic.  It would be hard to make guarantees about type safety otherwise.

> this would have issues with casting arbitrary memory to this type 

Correct. This is only safe to use on GC heap memory.